packages/eslint-plugin/src/rules/no-restricted-imports.ts

```diff
@@ -269,28 +269,55 @@ export default createRule<Options, MessageIds>({
       );
     }
 
-    return {
-      ImportDeclaration(node: TSESTree.ImportDeclaration): void {
+    function checkImportNode(node: TSESTree.ImportDeclaration): void {
+      if (
+        node.importKind === 'type' ||
+        (node.specifiers.length > 0 &&
+          node.specifiers.every(
+            specifier =>
+              specifier.type === AST_NODE_TYPES.ImportSpecifier &&
+              specifier.importKind === 'type',
+          ))
+      ) {
+        const importSource = node.source.value.trim();
         if (
-          node.importKind === 'type' ||
-          (node.specifiers.length > 0 &&
-            node.specifiers.every(
-              specifier =>
-                specifier.type === AST_NODE_TYPES.ImportSpecifier &&
-                specifier.importKind === 'type',
-            ))
+          !isAllowedTypeImportPath(importSource) &&
+          !isAllowedTypeImportPattern(importSource)
```

Comment:
[Cleanup] Nit: there's a teeny bit of duplication with calling `!isAllowedTypeImportPath(importSource) && !isAllowedTypeImportPattern(importSource)`. Maybe worth a helper function? It's only two places, so maybe not...